gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/IllegalMethodCalledDuringTaskExecution.java

```diff
@@ -55,7 +60,108 @@
         the tasks become independent and can execute in parallel.
         """)
 public final class IllegalMethodCalledDuringTaskExecution extends GradleGuideBugChecker
-        implements BugChecker.MethodTreeMatcher {
+        implements BugChecker.MethodInvocationTreeMatcher {
+    private static final LoadingCache<CompilationUnitTree, MethodCallGraph> callGraphCache =
+            Caffeine.newBuilder().maximumSize(1_000).weakKeys().softValues().build(MethodCallGraph::new);
+
+    private static final Matcher<ExpressionTree> PROJECT_GET_LOGGER = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.Project")
+            .named("getLogger");
+    private static final Matcher<ExpressionTree> TASK_GET_PROJECT = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.Task")
+            .named("getProject");
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        if (isSuppressed(tree, state)) {
+            return Description.NO_MATCH;
+        }
+
+        Optional<Violation> firstMatchingViolation = violations.stream()
+                .filter(violation -> violation.matches(tree, state))
+                .findFirst();
+        if (firstMatchingViolation.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+
+        CompilationUnitTree compilationUnit = state.getPath().getCompilationUnit();
+        MethodCallGraph callGraph = callGraphCache.get(compilationUnit);
+        MethodSymbol matched = ASTHelpers.getSymbol(tree);
+        Set<MethodSymbol> transitiveCallers = callGraph.transitiveCallers(matched);
```

Comment:
I wonder whether we should weakly cache this too - idk how expensive the search is, but if it's being called multiple times per method it might make sense.

Then again caching too much is also bad. I say we leave it for now but bear it mind if there is a performance issue with this check.